File: app/src/main/java/org/michaelbel/moviemade/rest/api/AUTHENTICATION.java
package org.michaelbel.moviemade.rest.api;

import org.michaelbel.moviemade.rest.model.Auth;

import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.Query;

@SuppressWarnings("all")
public interface AUTHENTICATION {

    @GET("authentication/token/new?")
    Call<Auth.CreatedRequest> createRequestToken(
            @Query("api_key") String apiKey
    );

    @GET("authentication/token/validate_with_login?")
    Call<Auth.ValidatedRequest> validateRequestToken(
            @Query("api_key") String apiKey,
            @Query("username") String username,
            @Query("password") String password,
            @Query("request_token") String requestToken
    );

    @GET("authentication/session/new?")
    Call<Auth.Session> createSession(
            @Query("api_key") String apiKey,
            @Query("request_token") String requestToken
    );

    @GET("authentication/guest_session/new?")
    Call<Auth.GuestSession> createGuestSession(
            @Query("api_key") String apiKey
    );
}

File: app/src/main/java/org/michaelbel/moviemade/data/service/AUTHENTICATION.kt
package org.michaelbel.moviemade.data.service

import io.reactivex.Observable
import org.michaelbel.moviemade.data.entity.*
import retrofit2.http.*

interface AUTHENTICATION {

    @GET("authentication/guest_session/new?")
    fun createGuestSession(
        @Query("api_key") apiKey: String
    ): Observable<GuestSession>

    @GET("authentication/token/new?")
    fun createRequestToken(
        @Query("api_key") apiKey: String
    ): Observable<Token>

    @POST("authentication/token/validate_with_login?")
    fun createSessionWithLogin(
        @Query("api_key") apiKey: String,
        @Body username: Username
    ): Observable<Token>

    @POST("authentication/session/new?")
    fun createSession(
        @Query("api_key") apiKey: String,
        @Body authToken: RequestToken
    ): Observable<Session>

    // createSession (from api v4)

    @HTTP(method = "DELETE", path = "authentication/session?", hasBody = true)
    fun deleteSession(
        @Query("api_key") apiKey: String,
        @Body sessionId: SessionId
    ): Observable<DeletedSession>
}